优化to_tensor函数中的bf16转换

Qin-sx opened the pull request:
<!-- TemplateReference: https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE -->
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->

### PR Category
<!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | CINN | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library | Auto Parallel | Inference | Environment Adaptation ] -->

User Experience

### PR Types
<!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Others ] -->

Improvements

### Description
<!-- Describe what you’ve done -->

resolve https://github.com/PaddlePaddle/Paddle/issues/72484

需求：

> 1.  保持原numpy类型不变，直接转为一个paddle.Tensor
> 2.  返回paddle.Tensor，astype为bfloat16

测试代码

```python
import numpy as np
import paddle
import time

x = np.random.randn(100000).astype(np.float32)
tensor_bfloat16 = paddle.to_tensor(x, dtype=paddle.bfloat16)


num_runs = 1000
total_time = 0

for _ in range(num_runs):
    start_time = time.time()
    tensor_bfloat16 = paddle.to_tensor(x, dtype=paddle.bfloat16)
    paddle.device.synchronize()
    end_time = time.time()
    total_time += (end_time - start_time)

avg_time = total_time / num_runs

print(f"avg time ({num_runs} runs): {avg_time * 1000:.4f} ms")
```

原本方式平均耗时为 40.4780 ms， 修改后平均耗时为 0.2655 ms。

Qin-sx committed:
optimized bf16 convert in to_tensor

	modified:   python/paddle/tensor/creation.py
	modified:   test/dygraph_to_static/test_to_tensor.py

SigureMo commented on test/dygraph_to_static/test_to_tensor.py:
1. 这段代码并没有修改静态图分支，不需要添加动转静单测，单测添加到 `test/legacy_test/test_eager_tensor.py`
2. 只是加了一个普通的函数跑不到，且添加 `unittest.skipIf` 的方式也是错的

Qin-sx replied:
```python
            tensor = core.eager.Tensor(
                value=data,
                place=place,
                persistable=False,
                zero_copy=False,
                name=None,
                stop_gradient=stop_gradient,
            )
            # tensor = tensor.astype(dtype) 
            tensor = paddle.cast(tensor, dtype)
            return tensor
```          

您好，可否帮忙看一下，在使用转换后，x.grad会变为None

```python
x = paddle.to_tensor( 1e6, dtype=paddle.bfloat16, stop_gradient=False)
print("x:", x)
y = x * x
y.backward()
print("x.grad:", x.grad)
```

修改后结果
```bash
x: Tensor(shape=[], dtype=bfloat16, place=Place(cpu), stop_gradient=False,
       -9.9942e+05)
x.grad: None
```
原本结果
```bash
x: Tensor(shape=[], dtype=bfloat16, place=Place(cpu), stop_gradient=False,
       -9.9942e+05)
x.grad: Tensor(shape=[], dtype=bfloat16, place=Place(cpu), stop_gradient=False,
       -1.9988e+06)
```

Qin-sx replied:
已修改

Qin-sx committed:
modified for grad

	modified:   python/paddle/tensor/creation.py
	modified:   test/dygraph_to_static/test_to_tensor.py

zhwesky2010 commented on python/paddle/tensor/creation.py:
```diff
@@ -763,7 +763,24 @@ def _handle_np_dtype(
+            tensor.stop_gradient = stop_gradient
+            return tensor
+        else:
+            data = _handle_np_dtype(data, dtype)
```
可以把_handle_np_dtype逻辑挪出来，移除掉原来不正确的bf16分支代码

Qin-sx replied:
请问这个是指将_handle_np_dtype函数删除，然后把代码放在else:中吗？因为_handle_np_dtype函数在前面也被调用过。

zhwesky2010 replied:
> 请问这个是指将_handle_np_dtype函数删除，然后把代码放在else:中吗？因为_handle_np_dtype函数在前面也被调用过。

这个逻辑很简单，可以都挪出来，直接处理掉原来错误的bf16分支

zhwesky2010 replied:
这个逻辑很简单，可以都挪出来，直接处理掉原来错误的bf16分支

Qin-sx replied:
嗯，收到，已修改

zhwesky2010 commented on python/paddle/tensor/creation.py:
```diff
@@ -763,7 +763,24 @@ def _handle_np_dtype(
+                name=None,
+                stop_gradient=stop_gradient,
+            )
+            tensor = tensor.detach().astype(dtype)
```
这样写呢：
```
tensor = core.eager.Tensor(
                value=data,
                place=place,
                persistable=False,
                zero_copy=False,
                name=None,
                stop_gradient=True,
)
tensor = tensor.astype('bfloat16')
tensor.stop_gradient = stop_gradient
return tensor
```

Qin-sx replied:
这样好像不行，上一个版本是类似的做法。我个人理解是只能数据进行转换，不能带着gradient一起转换。如果不用detach，会导致grad为None。

> ```python
>             tensor = core.eager.Tensor(
>                 value=data,
>                 place=place,
>                 persistable=False,
>                 zero_copy=False,
>                 name=None,
>                 stop_gradient=stop_gradient,
>             )
>             # tensor = tensor.astype(dtype) 
>             tensor = paddle.cast(tensor, dtype)
>             return tensor
> ```
> 
> 您好，可否帮忙看一下，在使用转换后，x.grad会变为None
> 
> ```python
> x = paddle.to_tensor( 1e6, dtype=paddle.bfloat16, stop_gradient=False)
> print("x:", x)
> y = x * x
> y.backward()
> print("x.grad:", x.grad)
> ```
> 
> 修改后结果
> 
> ```shell
> x: Tensor(shape=[], dtype=bfloat16, place=Place(cpu), stop_gradient=False,
>        -9.9942e+05)
> x.grad: None
> ```
> 
> 原本结果
> 
> ```shell
> x: Tensor(shape=[], dtype=bfloat16, place=Place(cpu), stop_gradient=False,
>        -9.9942e+05)
> x.grad: Tensor(shape=[], dtype=bfloat16, place=Place(cpu), stop_gradient=False,
>        -1.9988e+06)
> ```

zhwesky2010 replied:
> 这样好像不行，上一个版本是类似的做法。我个人理解是只能数据进行转换，不能带着gradient一起转换。如果不用detach，会导致grad为None。
> 
> > ```python
> >             tensor = core.eager.Tensor(
> >                 value=data,
> >                 place=place,
> >                 persistable=False,
> >                 zero_copy=False,
> >                 name=None,
> >                 stop_gradient=stop_gradient,
> >             )
> >             # tensor = tensor.astype(dtype) 
> >             tensor = paddle.cast(tensor, dtype)
> >             return tensor
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 您好，可否帮忙看一下，在使用转换后，x.grad会变为None
> > ```python
> > x = paddle.to_tensor( 1e6, dtype=paddle.bfloat16, stop_gradient=False)
> > print("x:", x)
> > y = x * x
> > y.backward()
> > print("x.grad:", x.grad)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 修改后结果
> > ```shell
> > x: Tensor(shape=[], dtype=bfloat16, place=Place(cpu), stop_gradient=False,
> >        -9.9942e+05)
> > x.grad: None
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 原本结果
> > ```shell
> > x: Tensor(shape=[], dtype=bfloat16, place=Place(cpu), stop_gradient=False,
> >        -9.9942e+05)
> > x.grad: Tensor(shape=[], dtype=bfloat16, place=Place(cpu), stop_gradient=False,
> >        -1.9988e+06)
> > ```

这样写是可以的，这是因为之前的写法，没有设置stop_gradient=True，使得第一个tensor与cast后的tensor建立了反向关系，梯度传导到第一个tensor了，而cast后的tensor的梯度就被清理掉了。

目前把第一个tensor的stop_gradient设置为True，就避免了这个问题。

Qin-sx replied:
嗯，收到，我测试了一下，确实没问题，感谢

Qin-sx committed:
changed core.eager.Tensor para

	modified:   python/paddle/tensor/creation.py

Qin-sx committed:
deleted _handle_np_dtype

	modified:   python/paddle/tensor/creation.py

zhwesky2010 commented on python/paddle/tensor/creation.py:
```diff
@@ -757,13 +742,35 @@ def _handle_np_dtype(
 
     if dtype:
-        data = _handle_np_dtype(data, dtype)
+        if (
```
这里不用搞这么多判断吧，代码注意**逻辑清晰，可读性强。**

这里要么全统一用 `convert_dtype(dtype)` 判断，要么统一用dtype判断，不用在这里反复冗余判断。ndarray也没必要判断吧

Qin-sx replied:
嗯，收到，已按照下面的方式修改

zhwesky2010 commented on python/paddle/tensor/creation.py:
```diff
@@ -757,13 +742,35 @@ def _handle_np_dtype(
+            tensor.stop_gradient = stop_gradient
+            return tensor
+        else:
+            if convert_dtype(dtype) != convert_dtype(data.dtype):
```
```
if dtype and  convert_dtype(dtype) != convert_dtype(data.dtype):
        if convert_dtype(dtype) == 'uint16':
             ...
        else:
             data = data.astype(convert_dtype(dtype))
```

这样可以吗，这里的分支显得又多又乱

Qin-sx replied:
嗯，收到，已修改，感谢

zhwesky2010 commented on python/paddle/tensor/creation.py:
```diff
@@ -757,13 +742,35 @@ def _handle_np_dtype(
                         else 'complex128'
                     )
-                data = _handle_np_dtype(data, default_type)
+                if convert_dtype(default_type) != convert_dtype(data.dtype):
```
这里统一写成：
```
if convert_dtype(default_type) != convert_dtype(data.dtype):
    dtype = default_type
```
然后交到下面的逻辑里去处理，代码更简洁

Qin-sx replied:
嗯，收到，已修改，感谢

zhwesky2010 reviewed (commented):
代码注意下可读性方面

Qin-sx committed:
updated conditions

	modified:   python/paddle/tensor/creation.py

zhwesky2010 reviewed (approved):
LGTM
